### DISCUSSION TITLE
spaceauth: Print FASTLANE_SESSION without formatting

### DISCUSSION THREAD

**Discussion by Semmu** [fastlane/fastlane#17843] [Ideas] [answered]
Hi everyone,

I'm using `fastlane spaceauth`'s output to provision MacOS VMs with Vagrant, which download, for example, Xcode using `xcversion`.

My current workflow is the following:
* Run a script on the host machine to authenticate to the Apple Developer Portal (_often requiring interaction because 2FA_) to get a valid `FASTLANE_SESSION` and export it to the current shell context.
* Run `vagrant up`, which in turn runs a provisioning script **inside** the VM with the required `FASTLANE_*` env. vars passed to the VM.
* The script provisions the VM: downloads Xcode and other stuff, etc.

Since `fastlane spaceauth` often requires interaction, I **have** to split the process into 2 scripts: one running on the host and one inside the VM (the provisioning script inside the VM must not require any interaction).

This solution works well, but I had to do some ugly hacks to get it working, since `spaceauth` outputs the session variable in a [formatted way](https://github.com/fastlane/fastlane/blob/48151291f2c4949c3b1b9919ba2cc81a7cc33293/spaceship/lib/spaceship/spaceauth_runner.rb#L61), which can't be easily eval'd.

My current solution looks something like this:

```bash
#!/bin/bash

# authenticate and write session to file
fastlane spaceauth 2>/dev/null | grep "export FASTLANE_SESSION" > session.txt

# remove fist 9 bytes (formatting)
dd if=session.txt of=session.trimmed.txt bs=1 skip=9 1>/dev/null

# cut last 9 bytes (reset formatting)
truncate --size=-9 session.trimmed.txt

eval $(cat session.trimmed.txt)
```

And you need to run it as `source ./apple_login.sh` to be able to export env. vars to the running shell context.

**So my question is:** could we get a command line flag to control the output of `fastlane spaceauth` so we can run it in a way which could be directly eval'd? It would eliminate the need for these ugly byte-trimming.

Something like:

```bash
eval $(fastlane spaceauth --eval)
```

What do you think?

**Comment by rogerluan** [comment]
Hey @Semmu 👋 

Could you try running `spaceauth` using [`FASTLANE_NO_COLORS=1`](https://docs.fastlane.tools/faqs/#disable-colored-output)? Something like:

```sh
FASTLANE_DISABLE_COLORS=1 fastlane spaceauth 2>/dev/null | grep "export FASTLANE_SESSION" > session.txt

// or

export FASTLANE_DISABLE_COLORS=1 
fastlane spaceauth 2>/dev/null | grep "export FASTLANE_SESSION" > session.txt
```

**Comment by Semmu** [comment]
Hi @rogerluan ,

Thanks for the suggestion! This solution indeed makes the process a bit simpler, since it eliminates the need for byte trimming, but it still feels a bit hacky and I would prefer some proper support for unattended scripts/usage.
I am willing to create a PR for it, but I want to hear the opinion of others as well.

**Comment by MarcPerezPro** [comment]
Thanks for the command line, we are now using
```bash
#!/usr/bin/env bash

bundle install
eval $(FASTLANE_DISABLE_COLORS=1 bundle exec fastlane spaceauth 2>/dev/null | grep "export FASTLANE_SESSION")
```
to set the `FASTLANE_SESSION` in our CI/CD.

We would love to do it that way:
```bash
eval $(fastlane spaceauth --eval)
```

**Reply by amirvenus** [reply]
I am using the same code but in my case `bundle install` was not necessary.

**Reply by rogerluan** [reply]
@return-main @amirvenus you can use this feature now https://github.com/fastlane/fastlane/pull/18538 - is that helpful?

**Reply by amirvenus** [reply]
> @return-main @amirvenus you can use this feature now https://github.com/fastlane/fastlane/pull/18538 - is that helpful?

Do you know if I can force Fastlane to ask for 2FA code i.e clearing the last logged in user completely?

**Reply by MarcPerezPro** [reply]
@amirvenus You can delete the cookie with `rm ~/.fastlane/spaceship/*/cookie`

**Comment by rogerluan** [comment, marked as answer]
I've opened a proposal for this feature here: https://github.com/fastlane/fastlane/pull/18538
cc @return-main @Semmu
